mining, ipc: omit dummy extraNonce from coinbase

Sjors opened the pull request:
This PR changes the Mining IPC interface to stop including a dummy `extraNonce` in the coinbase `scriptSig` by default, exposing only the consensus-required BIP34 height. This simplifies downstream mining software (including Stratum v2), avoids forcing clients to strip or ignore data we generate, and reduces the risk of incompatibilities if future soft forks add required commitments to the `scriptSig`.

Existing behavior is preserved for RPCs, tests, regtest, and internal mining by explicitly opting in to the dummy `extraNonce` where needed (e.g. to satisfy `bad-cb-length` at low heights), so consensus rules and test coverage are unchanged. The remainder of the PR consists of small comment fixes, naming clarifications, and test cleanups to make the intent and behavior clearer.

ajtowns commented on src/node/miner.cpp:
```diff
@@ -161,7 +161,12 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
-    coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
+    coinbaseTx.vin[0].scriptSig = CScript() << nHeight;
+    // IsTestChain() can be dropped if hardcoded block hashes in tests are regenerated
+    if (nHeight <= 16 || chainparams.IsTestChain()) {
```
Match the comment to the updated code too?

Sjors replied:
Fixed

shahsb reviewed (commented):
Concept ACK

ajtowns commented:
> Our miner code adds `OP_0` to the coinbase `scriptSig` in order to avoid triggering the `bad-cb-length` consensus error on test networks.

Correct me if I'm wrong, please! I think what's actually going on here is:

 * scriptSigs in the coinbase have had to be 2 bytes or more since before bitcoin's git history began
 * this is perhaps because mainnet bitcoin blocks will usually require some extranonce stuff, and the coinbase tx's scriptSig is a sensible place to do that
 * in any event, satoshi set the scriptSig up to contain a push of nBits and a push of the extranonce. this was changed to be time and extranonce (#505) and finally height and extranonce via BIP34 activation.
 * when BIP34 is applied to blocks 1-16, the coinbase is encoded as a 1-byte OP_1 through OP_16 (this is compliant with the "minimally encoded serialized CScript" part of the spec, though not the "first byte is number of bytes..." part).
 * this means the only blocks compliant with BIP34 that might potentially have a too short scriptSig are blocks 1 through 16
 * because mining regtest doesn't require extranonce stuff, there's no automatic reason to add extra bytes here, but the cb-length consensus check requires it anyway.
 * thus, when mining regtest blocks in the functional test framework, we explicitly add an OP_0 in `blocktools.py:script_BIP34_coinbase_height` for blocks 1 through 16. regtest was introduced a year after BIP 34, so there was never a question of "what did we do before BIP34 was enforced on regtest"
 * that code was introduced in #16363, at which point BIP34 wasn't activated in regtest until block height 500; this was lowered to 2 in #16333 at which point the code was used

So I think it's more fair to say that our miner code adds `OP_0` as a dummy extraNonce, expecting it to be incremented as header nonce space runs out, which just never happens in unit tests and regtest.

In the current code, having a dummy extraNonce actually seems sensible to me -- in real PoW contexts, we'd need a real extraNonce anyway, so this makes the test environment a little more similar to reality... So I feel a bit -0 on this as a consequence. I wonder if there isn't a way to have this work for stratumv2 without changing the way the existing code works?

Would it work for the stratumv2 interface (the part of it inside bitcoin core?) to just recognise we supply a dummy extra nonce, and drop it? Even for the first 16 blocks, sv2 miners that supply a non-empty extraNonce of their own, or that include a pool-signature in the coinbase will pass the cb-length check. And after the first 16 blocks, it's not an issue at all.

Sjors commented:
Rebased after #32155.

> So I think it's more fair to say that our miner code adds OP_0 as a dummy extraNonce

Possibly, but `extraNonce` seems like an implementation detail that should be left to miners. E.g. the Stratum v2 spec defines how to use it here: https://github.com/stratum-mining/sv2-spec/blob/main/05-Mining-Protocol.md

It doesn't belong in a block template imo.

> I wonder if there isn't a way to have this work for stratumv2 without changing the way the existing code works?

The [Template Distribution protocol](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#72-newtemplate-server---client) defines a message `NewTemplate` which has a `coinbase_prefix` field, described as:

> Up to 8 bytes (not including the length byte) which are to be placed at the beginning of the coinbase field in the coinbase transaction

https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#72-newtemplate-server---client

> just recognise we supply a dummy extra nonce, and drop it

Yes but this would be a foot gun if a future soft fork requires an additional commitment. It's also up to every consumer of our Mining interface to implement that (correctly), not just the one I wrote.

ajtowns commented:
> It doesn't belong in a block template imo.

Right, but we currently don't include it in a block template either, so that's (currently) fine.

> > Would it work for the stratumv2 interface ... to just recognise we supply a dummy extra nonce, and drop it
> 
> Yes but this would be a foot gun if a future soft fork requires an additional commitment. It's also up to every consumer of our Mining interface to implement that (correctly), not just the one I wrote.

I think it's more likely that any additional commitments required by future soft forks will be in the coinbase tx's outputs because the  coinbase scriptSig's limited to 100 bytes. The segwit commitment output is designed to allow for this, so additional outputs aren't needed; signet makes use of this ability for the block signature.

> We could implement it somewhere between the mining code and interface, so at least interface users don't have to deal with this. But currently `BlockAssembler::CreateNewBlock()` is called pretty much directly without further processing.

Yeah, this was what I was thinking. Maybe adding something like:

```c++
struct CBlockTemplate
{
    CBlock block;

    std::span<unsigned char> CoinbaseScriptSigPrefix()
    {
       if (block.vtx.size() == 0 || block.vtx[0].vin.size() == 0 || block.vtx[0].vin[0].scriptSig.size() == 0) return {};
       // our scriptSig includes a dummy extraNonce. Drop it here.
       std::span<unsigned char> r{block.vtx[0].vin[0].scriptSig};
       return r.first(r.size()-1);
    }
    ...
};
```

and whatever `getblocktemplate`-ish api we introduce for stratumv2 calls that function to provide the information rather than dumping the scriptSig directly.

ajtowns committed:
[doc] Update comments on dummy extraNonces in tests

Sjors commented:
I switched to the approach of adding an `include_dummy_extranonce`, having the tests set it to `true` while IPC calls use the default `false`. Expanded `test/functional/interface_ipc.py` to demonstrate it.

ryanofsky reviewed (approved):
Code review ACK b03feedadb411e15967659be83940d73ce977848. Assuming #33819 goes ahead, there should be less need for this change from the mining interface perspective. But not adding OP_0 seems like more sensible default behavior and all the changes here seem nice to make the test code and BlockAssembler code more explicit and clear about what they are doing.

ryanofsky reviewed (approved):
Code review ACK 9832a9aa2bffb5c0e14eb67ce22d9ce900106304. Just rebased since last review and tweaked comment

ryanofsky reviewed (approved):
Code review ACK a3e887766459f850a9ff6f7c84139baa63da50c0, just fixing conflict in test since last review

Sjors committed:
[miner] omit dummy extraNonce via IPC

Previously the coinbase transaction generated by our miner code was
not used downstream, because the getblocktemplate RPC excludes it.

Since the Mining IPC interface was introduced in #30200 we do expose
this dummy coinbase transaction. In Stratum v2 several parts of it
are communicated downstream, including the scriptSig.

This commit removes the dummy extraNonce from the coinbase scriptSig
in block templates requested via IPC. This limits the scriptSig
to what is essential for consensus (BIP34) and removes the need for
external mining software to remove the dummy, or even ignore
the scriptSig we provide and generate it some other way. This
could cause problems if a future soft fork requires additional
data to be committed here.

A test is added to verify the new IPC behavior.

It achieves this by introducing an include_dummy_extranonce
option which defaults to false with all test code updated to
set it to true. Because this option is not exposed via IPC,
callers will no longer see it.

The caller needs to ensure that for blocks 1 through 16
they pad the scriptSig in order to avoid bad-cb-length.

Co-authored-by: Anthony Towns <aj@erisian.com.au>

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -177,11 +177,15 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
+    coinbaseTx.vin[0].scriptSig = CScript() << nHeight;
+    if (m_options.include_dummy_extranonce) {
+        // Add trailing OP_0 (historically an extranonce). It's not consensus
+        // required, but it prevents bad-cb-length for nHeight <= 16.
```
In commit "[miner] omit dummy extraNonce via IPC" (75bce1e711351e2d3e512c4a24db0275afe5a999)

Would be helpful to add "in tests" to the end of this sentence, or clarify where else the bad-cb-length

Sjors replied:
Not just in tests, also on new signets. And since we removed checkpoints, any aliens using mining code for a giant reorg back to genesis also need to be careful :-)

ryanofsky reviewed (approved):
Code review ACK 75bce1e711351e2d3e512c4a24db0275afe5a999. Was rebased to avoid conflicts with https://github.com/bitcoin/bitcoin/pull/33819 since last review

Sjors committed:
test: clarify getCoinbaseRawTx() comparison

The code comment mistakingly referred to "the deprecated getCoinbaseTx()",
instead of getCoinbaseRawTx. This was missed in d59b4cdb5772917ee13e48552d51662160104b62.

Also rename parse_and_deserialize_coinbase_tx to make it more clear
it refers to the deprecated method.

Finally, this commit drops the getCoinbaseRawTx() call when testing
template inspectors. The coinbase input check here is already covered by
build_coinbase_test.

ryanofsky reviewed (approved):
Code review ACK d511adb664edcfb97be44bc0738f49b679240504. Just rebased since last review and make suggested tweaks. I'd really like to see this PR merged for the cleanups and sanity it brings to this code. Needs another reviewer though.

I feel like current PR description may be a hurdle to getting this reviewed since it is very technical and immediately goes into weeds with details about the history of the code and about stratum v2. All this information is useful but IMO it would be much better to put in the third commit message or in the IPC interface documentation where it will be more accessible anyway after this is merged. Then the PR description can just say why this change is good and what information is most useful to review it. Chatgpt suggested following which reads very well to me:

> This PR changes the Mining IPC interface to stop including a dummy extraNonce in the coinbase scriptSig by default, exposing only the consensus-required BIP34 height. This simplifies downstream mining software (including Stratum v2), avoids forcing clients to strip or ignore data we generate, and reduces the risk of incompatibilities if future soft forks add required commitments to the scriptSig.
>
> Existing behavior is preserved for RPCs, tests, regtest, and internal mining by explicitly opting in to the dummy extraNonce where needed (e.g. to satisfy bad-cb-length at low heights), so consensus rules and test coverage are unchanged. The remainder of the PR consists of small comment fixes, naming clarifications, and test cleanups to make the intent and behavior clearer.

It also suggested some shorter titles which may be good

- mining ipc: omit dummy extraNonce from coinbase
- mining ipc: omit dummy extraNonce (tests and RPC unchanged)
- mining: make IPC coinbase scriptSig future-proof

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -144,7 +144,7 @@ async def parse_and_deserialize_block(self, block_template, ctx):
         return block
 
-    async def parse_and_deserialize_coinbase_tx(self, block_template, ctx):
+    async def get_coinbase_raw_tx(self, block_template, ctx):
```
I'll rename `parse_and_deserialize_coinbase` if I need to retouch, but will leave `parse_and_deserialize_block` as is for now, to not expand the scope of this PR.

sedited commented on src/node/types.h:
```diff
@@ -67,6 +67,10 @@ struct BlockCreateOptions {
+    /**
+     * Whether to include an OP_0 as a dummy extraNonce in the template's coinbase
+     */
+    bool include_dummy_extranonce{false};
```
In commit d511adb664edcfb97be44bc0738f49b679240504

Seems like this commit could be de-risked if this were true instead, but I guess you didn't do that to discourage it in the future?

sedited reviewed (approved):
ACK d511adb664edcfb97be44bc0738f49b679240504

achow101 commented:
ACK d511adb664edcfb97be44bc0738f49b679240504
